packages/components/src/components/Log/Log.js

```diff
@@ -28,17 +30,147 @@ const itemSize = 15; // This should be kept in sync with the line-height in SCSS
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    if (
+      prevState.logs?.length !== this.state.logs?.length ||
+      prevProps.isLogsMaximized !== this.props.isLogsMaximized
+    ) {
+      if (this.shouldAutoScroll(this.props.stepStatus, prevProps.stepStatus)) {
+        this.scrollToBottomLog();
+        return; // the scroll event listener will call handleLogScroll
+      }
+      this.handleLogScroll();
+    }
   }
 
   componentWillUnmount() {
     clearInterval(this.timer);
+    window.removeEventListener("scroll", this.handleLogScroll, true);
     this.cancelled = true;
   }
 
+  handleLogScroll = () => {
+    this.updateScrollButtonHorizontalCoordinate();
+    const isLogTopUnseen = this.isLogTopUnseen();
+    const isLogBottomUnseen = this.isLogBottomUnseen();
+
+    if (
+      isLogBottomUnseen !== this.state.isLogBottomUnseen ||
+      isLogTopUnseen !== this.state.isLogTopUnseen
+    ) {
+      this.setState({
+        isLogBottomUnseen,
+        isLogTopUnseen
+      });
+    }
+  };
+  
+  isLogBottomUnseen = () => {
+    return utils.isElementEndBelowViewBottom(this.logRef?.current)
+      || utils.hasElementPositiveScrollBottom(this.textRef?.current?.firstElementChild);
+  };
+
+  isLogTopUnseen = () => {
+    return utils.isElementStartAboveViewTop(this.logRef?.current)
+      || utils.hasElementPositiveScrollTop(this.textRef?.current?.firstElementChild);
+  };
+
+  scrollToBottomLog = () => {
+    const longTextElement = this.textRef?.current?.firstElementChild;
+    if (utils.hasElementPositiveScrollBottom(longTextElement)) {
+      longTextElement.scrollTop = longTextElement.scrollHeight - longTextElement.clientHeight;
+    }
+    const rootElement = document.documentElement;
+    rootElement.scrollTop = rootElement.scrollHeight - rootElement.clientHeight;
+  };
+
+  scrollToTopLog = () => {
+    const longTextElement = this.textRef?.current?.firstElementChild;
+    if (utils.hasElementPositiveScrollTop(longTextElement)) {
+      longTextElement.scrollTop = 0;
+    }
+    document.documentElement.scrollTop = 0;
+  };
+
+  shouldAutoScroll = (stepStatus, prevStepStatus) => {
+    return this.props.enableLogAutoScroll && !this.state.isLogBottomUnseen &&
+      (
+        (stepStatus && !stepStatus.terminated) ||
+        (stepStatus?.terminated && prevStepStatus && !prevStepStatus.terminated)
+      ) &&
+      this.isLogBottomUnseen();
+  };
+
+  updateScrollButtonHorizontalCoordinate = () => {
+    const logElementRight = (window.innerWidth || document.documentElement.clientWidth) -
+      this.logRef?.current?.getBoundingClientRect().right;
+    document.documentElement.style.setProperty("--log-element-right",
+      `${logElementRight.toString()}px`);
+  };
+
+  getScrollButtons = () => {
+    const { enableLogScrollButtons, intl } = this.props;
+    if (!enableLogScrollButtons) {
+      return <Fragment />;
+    }
+
+    return (
+      <Fragment>
+        {this.state.isLogTopUnseen ? (
+          <button
```

Comment:
Can we use a Carbon `Button` component here? This will ensure consistent styling across future updates.

Reply:
I used the `button` tag because it is already used in https://github.com/tektoncd/dashboard/blob/aec426b0c0d7004dcff1dec627be6ee004b9ea37/packages/components/src/components/LogsToolbar/LogsToolbar.js#L27